File: src/implementation/sheet/sheet.rs
use std::{collections::HashMap, error::Error};

use crate::{
    api::db::{IntoTable, Table, TableQuery, TableRow},
    implementation::drive::google_drive_file::GoogleDriveFile,
    FileMetadata,
};

use super::{get, save, Sheet};

impl IntoTable<Sheet> for GoogleDriveFile {
    fn into_table(&self) -> Sheet {
        Sheet {
            session: self.get_session(),
            spreadsheet_id: self.get_id(),
        }
    }
}
fn transform(
    row: &TableRow<u64>,
    column_map: &HashMap<String, usize>,
) -> (Vec<String>, HashMap<String, usize>) {
    let mut column_map = column_map.clone();
    let mut data = row.get_data().clone();
    let mut existing_rows_list = column_map.iter().fold(
        vec![String::new(); column_map.len()],
        |mut acc, (key, value)| {
            acc[value.clone()] = match data.remove(key) {
                Some(value) => value,
                None => "".into(),
            };
            acc
        },
    );
    row.get_data().iter().for_each(|(key, value)| {
        if !column_map.contains_key(&key.to_string()) {
            column_map.insert(key.to_string(), column_map.len());
            existing_rows_list.push(value.clone());
        }
    });
    return (existing_rows_list, column_map);
}
impl Table for Sheet {
    type IdType = u64;
    fn save_all(&self, data: Vec<TableRow<Self::IdType>>) -> Result<(), Box<dyn Error>> {
        let columns = get::columns(self)?;

        let mut column_map = columns.iter().enumerate().fold(
            HashMap::<String, usize>::new(),
            |mut acc, (index, value)| {
                acc.insert(value.clone(), index);
                acc
            },
        );
        let clean_rows = data.iter().filter(|row| row.get_id() == &None).fold(
            Vec::<Vec<String>>::with_capacity(data.len()),
            |mut acc, row| {
                let (row_list, new_column_map) = transform(row, &column_map);
                column_map = new_column_map;
                acc.push(row_list);
                acc
            },
        );
        let mut persisted_rows = data.iter().filter(|row| row.get_id() != &None).fold(
            Vec::<(Self::IdType, Vec<String>)>::with_capacity(data.len()),
            |mut acc, row| {
                let (row_list, new_column_map) = transform(row, &column_map);
                column_map = new_column_map;
                acc.push((row.get_id().unwrap(), row_list));
                acc
            },
        );
        let column_row = column_map.iter().fold(
            vec![String::new(); column_map.len()],
            |mut acc, (key, value)| {
                acc[value.clone()] = key.clone();
                acc
            },
        );
        persisted_rows.push((0, column_row));
        save::persisted_rows(self, persisted_rows)?;
        save::new_rows(self, clean_rows)?;
        Ok(())
    }

    fn find_by_query(
        &self,
        query: TableQuery,
    ) -> Result<Vec<TableRow<Self::IdType>>, Box<dyn Error>> {
        let columns = get::columns(&self)?;
        let data = get::rows(&self, query)?;
        let data = data
            .iter()
            .map(|(id, row)| {
                let map = columns.iter().enumerate().fold(
                    HashMap::<String, String>::new(),
                    |mut acc, (index, key)| {
                        let value = row.get(index);
                        if value.is_none() {
                            return acc;
                        }

                        acc.insert(key.to_string(), value.unwrap().to_string());
                        acc
                    },
                );
                TableRow::new_persisted(id.clone(), map)
            })
            .collect();

        Ok(data)
    }
}

File: src/api/implementation/sheet/sheet.ts
import { Table } from "../../db/Table";
import { TableQuery } from "../../db/TableQuery";
import { TableQueryBuilder } from "../../db/TableQueryBuilder";
import { TableRow } from "../../db/TableRow";
import { GoogleSession } from "../GoogleSession";
import { getColumns } from "./columns";
import { newRows } from "./newRows";
import { persistedRows } from "./persistedRows";
import { getRows } from "./rows";

export class Sheet {
  session: GoogleSession; // Replace 'any' with the appropriate session type
  spreadsheet_id: string;

  constructor(session: any, spreadsheet_id: string) {
    this.session = session;
    this.spreadsheet_id = spreadsheet_id;
  }
}

export type ColumnMap = {
  [key: string]: number;
}

export class SheetTable implements Table<number> {
  idType: number;
  sheet: Sheet;

  constructor(sheet: Sheet) {
    this.idType = 0; // Replace with the appropriate IdType
    this.sheet = sheet;
  }

  async save_all(data: TableRow<number>[]): Promise<void> {
    const columns = await getColumns(this.sheet);
    let column_map: ColumnMap = columns.reduce((acc: ColumnMap, value: string, index: number) => {
      acc[value] = index;
      return acc;
    }, {});

    const clean_rows = data.reduce((acc: string[][], row: TableRow<number>) => {
      if (!row.id) {
        const [row_list, new_column_map] = transform(row, column_map);
        column_map = new_column_map;
        acc.push(row_list);
      }
      return acc;
    }, []);

    const persisted_rows = data.reduce((acc: [number, string[]][], row: TableRow<number>) => {
      if (row.id) {
        const [row_list, new_column_map] = transform(row, column_map);
        column_map = new_column_map;
        acc.push([row.id, row_list]);
      }
      return acc;
    }, []);

    const column_row = Object.entries(column_map).reduce(
      (acc: string[], [key, value]: [string, number]) => {
        acc[value] = key;
        return acc;
      },
      Array(Object.keys(column_map).length).fill('')
    );

    persisted_rows.push([0, column_row]);

    await persistedRows(this.sheet, persisted_rows);
    await newRows(this.sheet, clean_rows);
  }

  async find_by_query(query: TableQuery): Promise<TableRow<number>[]> {
    const columns = await getColumns(this.sheet);
    const data = await getRows(this.sheet, query);

    const parsedData = data.map(([id, row]: [number, string[]]) => {
      const map = row.reduce((acc: { [key: string]: string }, value, index) => {
        const key = columns[index];
        if (value) {
          acc[key] = value;
        }
        return acc;
      }, {});
      return {
        id: id,
        data: map
      } satisfies TableRow<number>;
    });

    return parsedData;
  }

  find_by(): TableQueryBuilder<number> {
    return new TableQueryBuilder(this);
  }
}

function transform(row: TableRow<number>, column_map: ColumnMap): [string[], ColumnMap] {
  const existing_rows_list: string[] = Array(column_map.length).fill('');

  const data = { ...row.data };
  for (const [key, value] of Object.entries(data)) {
    if (column_map[key] === undefined) {
      column_map[key] = Object.keys(column_map).length;
      existing_rows_list.push(value);
    } else {
      existing_rows_list[column_map[key]] = value;
      delete data[key];
    }
  }

  return [existing_rows_list, column_map];
}